### PR TITLE
[SPARK-15495][SQL]Improve the explain output for Aggregation operator

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

This PR improves the explain output of Aggregator operator. 

SQL:

```
Seq((1,2,3)).toDF("a", "b", "c").createTempView("df1")
spark.sql("cache table df1")
spark.sql("select count(a), count(c), b from df1 group by b").explain()
```

**Before change:**

```
*TungstenAggregate(key=[b#8], functions=[count(1),count(1)], output=[count(a)#79L,count(c)#80L,b#8])
+- Exchange hashpartitioning(b#8, 200), None
   +- *TungstenAggregate(key=[b#8], functions=[partial_count(1),partial_count(1)], output=[b#8,count#98L,count#99L])
      +- InMemoryTableScan [b#8], InMemoryRelation [a#7,b#8,c#9], true, 10000, StorageLevel(disk=true, memory=true, offheap=false, deserialized=true, replication=1), LocalTableScan [a#7,b#8,c#9], [[1,2,3]], Some(df1)
```

**After change:**

```
*Aggregate(key=[b#8], functions=[count(1),count(1)], output=[count(a)#79L,count(c)#80L,b#8])
+- Exchange hashpartitioning(b#8, 200), None
   +- *Aggregate(key=[b#8], functions=[partial_count(1),partial_count(1)], output=[b#8,count#98L,count#99L])
      +- InMemoryTableScan [b#8], InMemoryRelation [a#7,b#8,c#9], true, 10000, StorageLevel(disk, memory, deserialized, 1 replicas), LocalTableScan [a#7,b#8,c#9], [[1,2,3]], Some(df1)
```
## How was this patch tested?

Manual test and existing UT.

### CODE REVIEW

**Comment by clockfly** [comment]
Some useful improvements in PR https://github.com/apache/spark/pull/13271 will be converted in new PRs.

File: core/src/main/scala/org/apache/spark/storage/StorageLevel.scala [hunk 1]
```diff
@@ -120,8 +120,14 @@ class StorageLevel private(
...
   override def toString: String = {
-    s"StorageLevel(disk=$useDisk, memory=$useMemory, offheap=$useOffHeap, " +
-      s"deserialized=$deserialized, replication=$replication)"
+    val disk = if (useDisk) "disk" else ""
+    val memory = if (useMemory) "memory" else ""
+    val heap = if (useOffHeap) "offheap" else ""
+    val deserialize = if (deserialized) "deserialized" else ""
+
+    val output =
+      Seq(disk, memory, heap, deserialize, s"$replication replications").filter(_.nonEmpty)
```
Comment (liancheng) [inline core/src/main/scala/org/apache/spark/storage/StorageLevel.scala]: Nit: "replications" => "replicas"?

**Comment by cloud-fan** [comment]
Can you add some test case to show the new format?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala [hunk 2]
```diff
@@ -126,7 +126,13 @@ private[sql] case class AggregateExpression(
     AttributeSet(childReferences)
   }
 
-  override def toString: String = s"($aggregateFunction,mode=$mode,isDistinct=$isDistinct)"
+  override def toString: String = {
+    if (isDistinct) {
+      s"$aggregateFunction(distinct, $mode)"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala]: this is different from what we discussed before, we need something like `partitial_sum(distinct a)`, instead of `sum(a)(distinct, partitial)`

**Comment by clockfly** [comment]
@cloud-fan  updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala [hunk 3]
```diff
@@ -126,7 +126,14 @@ private[sql] case class AggregateExpression(
...
 
-  override def toString: String = s"($aggregateFunction,mode=$mode,isDistinct=$isDistinct)"
+  override def toString: String = {
+    val agg = aggregateFunction
+    if (isDistinct) {
+      s"${mode}_${agg.prettyName}(distinct ${agg.argString(false)})"
+    } else {
+      s"${mode}_${agg.simpleString}"
+    }
+  }
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala]: Personally I'd prefer `:` instead of `_` as separator between `mode` and `agg`. Because `_` is a valid identifier character, while `:` is not, thus it's more obvious that `partial:` is a prefix rather part of the aggregate function name.

(But that's pretty subjective.)

**Comment by liancheng** [comment]
Seems that the examples listed in the PR description is inconsistent with the actual changes? For example, PR description says `count(1)(Partial)` while the actual changes suggest something like `File_count(1)`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 4]
```diff
@@ -424,23 +424,50 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
...
-  def simpleString: String = s"$nodeName $argString".trim
+  /**
+   * Returns a string representing the arguments to this node
+   *
+   * @param verbose Controls whether to output more verbose string or not.
+   */
+  def argString(verbose: Boolean): String = {
+    productIterator.flatMap {
+      case tn: TreeNode[_] if containsChild(tn) => Nil
+      case tn: TreeNode[_] => s"${if (verbose) tn.verboseString else tn.simpleString}" :: Nil
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: Nit: You can simply move the `if` expression out of `s"..."`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 5]
```diff
@@ -424,23 +424,50 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
...
+      case seq: Seq[BaseType] if seq.toSet.subsetOf(children.toSet) => Nil
+      case seq: Seq[_] =>
+        val output = seq.map {
+          case tn: TreeNode[_] =>
+            s"${if (verbose) tn.verboseString else tn.simpleString}"
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: Same as above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 6]
```diff
@@ -424,23 +424,50 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
...
+  /**
+   * Returns a string representing the arguments to this node
+   *
+   * @param verbose Controls whether to output more verbose string or not.
+   */
+  def argString(verbose: Boolean): String = {
+    productIterator.flatMap {
+      case tn: TreeNode[_] if containsChild(tn) => Nil
+      case tn: TreeNode[_] => s"${if (verbose) tn.verboseString else tn.simpleString}" :: Nil
+      case seq: Seq[BaseType] if seq.toSet.subsetOf(children.toSet) => Nil
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: I believe `BaseType` here is not checked at compile time and results in a compile time warning.

Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: I copied it from old code. I also think it is buggy. Let me fix it.

Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: Well, it's not a bug, but just not checked as expected...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 7]
```diff
@@ -424,23 +424,50 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
...
+      case iter: Iterable[_] if iter.isEmpty => Nil
+      case set: Set[_] => set.mkString("{", ", ", "}") :: Nil
+      case array: Array[_] => array.mkString("[", ", ", "]") :: Nil
+      case null => Nil
+      case false => Nil
+      case None => Nil
+      case Some(null) => Nil
+      case Some(str: String) => str :: Nil
+      case other => other :: Nil
+    }.mkString(", ").trim()
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: Why `.trim()` here? (BTW, `.trim` instead of `.trim()`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 8]
```diff
@@ -201,24 +201,26 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
...
+      stringOrError(output),
+      stringOrError(analyzed.treeString(verbose = true))
+    ).filter(_.nonEmpty).mkString("\n")
 
     s"""== Parsed Logical Plan ==
-       |${stringOrError(logical)}
+       |${stringOrError(logical.treeString(verbose = true))}
        |== Analyzed Logical Plan ==
-       |$analyzedPlan
+       |${analyzedPlan}
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Nit: Revert this change.

**Comment by rxin** [comment]
I submitted a small pr here: https://github.com/apache/spark/pull/13367

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortBasedAggregateExec.scala [hunk 9]
```diff
@@ -101,11 +101,23 @@ case class SortBasedAggregateExec(
...
     val allAggregateExpressions = aggregateExpressions
 
     val keyString = groupingExpressions.mkString("[", ",", "]")
     val functionString = allAggregateExpressions.mkString("[", ",", "]")
     val outputString = output.mkString("[", ",", "]")
-    s"SortBasedAggregate(key=$keyString, functions=$functionString, output=$outputString)"
+    if (verbose) {
+      s"SortAggregate(key=$keyString, functions=$functionString, output=$outputString)"
+    } else {
+      s"SortAggregate(key=$keyString, functions=$functionString"
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortBasedAggregateExec.scala]: Missing `)` in the output string.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala [hunk 10]
```diff
@@ -760,16 +760,29 @@ case class TungstenAggregate(
...
-        s"TungstenAggregate(key=$keyString, functions=$functionString, output=$outputString)"
+
+        if (verbose) {
+          s"Aggregate(key=$keyString, functions=$functionString, output=$outputString)"
+        } else {
+          s"Aggregate(key=$keyString, functions=$functionString)"
+        }
       case Some(fallbackStartsAt) =>
-        s"TungstenAggregateWithControlledFallback $groupingExpressions " +
+        s"AggregateWithControlledFallback $groupingExpressions " +
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala]: Not quite sure about this one, but is this change intentional?

**Comment by clockfly** [comment]
retest this please

**Comment by clockfly** [comment]
retest this please

**Comment by rxin** [comment]
LGTM pending Jenkins.

**Comment by cloud-fan** [comment]
thanks, merging to master and 2.0
